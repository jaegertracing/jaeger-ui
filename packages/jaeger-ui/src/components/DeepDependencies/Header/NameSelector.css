/*
Copyright (c) 2019 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.NameSelector {
  border-bottom: 1px solid transparent;
  color: var(--tx-color-muted);
  cursor: pointer;
  font-size: 1.5em;
  margin: 0 1.3em 0 0;
  outline: 4px solid transparent;
}

.NameSelector.is-invalid {
  background: #fff2e888;
  border-color: #ffbb96;
  outline-color: #fff2e888;
}

.NameSelector.is-invalid:hover {
  background: #fff2e8;
  border-color: #ff9c6e;
  outline-color: #fff2e8;
}

.NameSelector:hover {
  background: #f4f4f4;
  border-color: #ccc;
  color: var(--tx-color-body);
  outline-color: #f4f4f4;
}

.NameSelector.is-active {
  background: #ddd;
  border-color: #bbb;
  color: var(--tx-color-body);
  outline-color: #ddd;
}

.NameSelector--label {
  font-weight: 400;
  padding-right: 0.5em;
}

.NameSelector--value {
  color: var(--tx-color-title);
}

.NameSelector--chevron {
  font-size: 0.6em;
  margin-left: 0.8em;
  position: relative;
}

.NameSelector--overlay .ant-popover-arrow {
  border: 1px solid #999;
  z-index: -1;
  background: #bbb;
}

.NameSelector--overlay .ant-popover-inner {
  border: 1px solid #ccc;
  overflow: hidden;
}
