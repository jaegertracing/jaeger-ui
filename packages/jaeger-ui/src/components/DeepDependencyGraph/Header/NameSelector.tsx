// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Popover } from 'antd';
import cx from 'classnames';
import IoChevronDown from 'react-icons/lib/io/chevron-down';

import BreakableText from '../../common/BreakableText';
import FilteredList from '../../common/FilteredList';

import './NameSelector.css';

type TProps = {
  label: string;
  placeholder?: boolean | string;
  options: string[];
  value: string | null;
  required?: boolean;
  setValue: (value: string) => void;
};

type TState = {
  popoverVisible: boolean;
};

export const DEFAULT_PLACEHOLDER = 'Select a valueâ€¦';

export default class NameSelector extends React.PureComponent<TProps, TState> {
  listRef: React.RefObject<FilteredList> = React.createRef();
  state: TState = { popoverVisible: false };

  componentDidUpdate() {
    if (this.listRef.current) {
      this.listRef.current.focusInput();
    }
  }

  setValue = (value: string) => {
    this.props.setValue(value);
    this.setState({ popoverVisible: false });
  };

  onFilterCancelled = () => {
    this.setState({ popoverVisible: false });
  };

  onPopoverVisbileChanged = (popoverVisible: boolean) => {
    this.setState({ popoverVisible });
  };

  render() {
    const { label, placeholder = false, options, required = false, value } = this.props;
    const { popoverVisible } = this.state;

    const rootCls = cx('NameSelector', {
      'is-active': popoverVisible,
      'is-invalid': required && !value,
    });
    let useLabel = true;
    let text = value || '';
    if (!value && placeholder) {
      useLabel = false;
      text = typeof placeholder === 'string' ? placeholder : DEFAULT_PLACEHOLDER;
    }
    return (
      <Popover
        overlayClassName="NameSelector--overlay u-rm-popover-content-padding"
        onVisibleChange={this.onPopoverVisbileChanged}
        placement="bottomLeft"
        content={
          <FilteredList
            ref={this.listRef}
            cancel={this.onFilterCancelled}
            options={options}
            value={value}
            setValue={this.setValue}
          />
        }
        trigger="click"
        visible={popoverVisible}
      >
        <h2 className={rootCls}>
          {useLabel && <span className="NameSelector--label">{label}</span>}
          <BreakableText className="NameSelector--value" text={text} />
          <IoChevronDown className="NameSelector--chevron" />
        </h2>
      </Popover>
    );
  }
}
