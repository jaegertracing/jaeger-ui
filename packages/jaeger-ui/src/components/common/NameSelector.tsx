// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Icon, Popover } from 'antd';
import cx from 'classnames';
import IoChevronDown from 'react-icons/lib/io/chevron-down';

import BreakableText from './BreakableText';
import FilteredList from './FilteredList';

import './NameSelector.css';

type TOptional = {
  clearValue: () => void;
  required?: false;
};

type TRequired = {
  clearValue?: never;
  required: true;
};

type TProps = {
  label: string;
  placeholder?: boolean | string;
  options: string[];
  value: string | null;
  setValue: (value: string) => void;
} & (TOptional | TRequired);

type TState = {
  popoverVisible: boolean;
};

export const DEFAULT_PLACEHOLDER = 'Select a valueâ€¦';

export default class NameSelector extends React.PureComponent<TProps, TState> {
  listRef: React.RefObject<FilteredList> = React.createRef();
  state: TState = { popoverVisible: false };

  componentDidUpdate() {
    if (this.listRef.current && this.state.popoverVisible) {
      this.listRef.current.focusInput();
    }
  }

  private changeVisible(popoverVisible: boolean) {
    if (popoverVisible) {
      window.document.body.addEventListener('click', this.onBodyClicked);
    } else {
      window.document.body.removeEventListener('click', this.onBodyClicked);
    }
    this.setState({ popoverVisible });
  }

  private clearValue = (evt: React.MouseEvent<React.ReactSVGElement>) => {
    if (this.props.required) throw new Error('Cannot clear value of required NameSelector');

    evt.stopPropagation();
    this.props.clearValue();
  };

  setValue = (value: string) => {
    this.props.setValue(value);
    this.changeVisible(false);
  };

  private onBodyClicked = () => {
    if (this.listRef.current && !this.listRef.current.isMouseWithin()) {
      this.changeVisible(false);
    }
  };

  onFilterCancelled = () => {
    this.changeVisible(false);
  };

  onPopoverVisibleChanged = (popoverVisible: boolean) => {
    this.changeVisible(popoverVisible);
  };

  render() {
    const { label, options, placeholder = false, required = false, value } = this.props;
    const { popoverVisible } = this.state;

    const rootCls = cx('NameSelector', {
      'is-active': popoverVisible,
      'is-invalid': required && !value,
    });
    let useLabel = true;
    let text = value || '';
    if (!value && placeholder) {
      useLabel = false;
      text = typeof placeholder === 'string' ? placeholder : DEFAULT_PLACEHOLDER;
    }
    return (
      <Popover
        overlayClassName="NameSelector--overlay u-rm-popover-content-padding"
        onVisibleChange={this.onPopoverVisibleChanged}
        placement="bottomLeft"
        content={
          <FilteredList
            ref={this.listRef}
            cancel={this.onFilterCancelled}
            options={options}
            value={value}
            setValue={this.setValue}
          />
        }
        trigger="click"
        visible={popoverVisible}
      >
        <h2 className={rootCls}>
          {useLabel && <span className="NameSelector--label">{label}:</span>}
          <BreakableText className="NameSelector--value" text={text} />
          <IoChevronDown className="NameSelector--chevron" />
          {!required && value && (
            <Icon className="NameSelector--clearIcon" type="close" onClick={this.clearValue} />
          )}
        </h2>
      </Popover>
    );
  }
}
