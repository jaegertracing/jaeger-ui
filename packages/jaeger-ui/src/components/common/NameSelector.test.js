// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { Icon, Popover } from 'antd';
import { shallow } from 'enzyme';

import BreakableText from './BreakableText';
import NameSelector, { DEFAULT_PLACEHOLDER } from './NameSelector';

describe('<NameSelector>', () => {
  const placeholder = 'This is the placeholder';
  let props;
  let wrapper;

  beforeEach(() => {
    props = {
      placeholder,
      label: 'a-label',
      options: ['a', 'b', 'c'],
      value: null,
      required: true,
      setValue: jest.fn(),
    };
    wrapper = shallow(<NameSelector {...props} />);
  });

  it('renders without exploding', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with is-invalid when required and without a value', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders without is-invalid when not required and without a value', () => {
    wrapper.setProps({ required: undefined });
    expect(wrapper).toMatchSnapshot();
  });

  describe('placeholder prop', () => {
    it('renders the placeholder when it is a string and value == null', () => {
      wrapper.setProps({ placeholder, value: null });
      expect(wrapper.find(BreakableText).prop('text')).toBe(placeholder);
    });

    it('renders the default placeholder when the prop is true and value == null', () => {
      wrapper.setProps({ placeholder: true, value: null });
      expect(wrapper.find(BreakableText).prop('text')).toBe(DEFAULT_PLACEHOLDER);
    });

    it('does not render a placeholder if there is a value', () => {
      const value = 'some-value';
      wrapper.setProps({ placeholder, value });
      expect(wrapper.find(BreakableText).prop('text')).toBe(value);
      wrapper.setProps({ placeholder: true, value });
      expect(wrapper.find(BreakableText).prop('text')).toBe(value);
    });

    it('does not render default placeholder if placeholder is disabled', () => {
      wrapper.setProps({ placeholder: undefined });
      expect(wrapper.find(BreakableText).prop('text')).toBe('');
    });
  });

  it('allows the filtered list to set values', () => {
    const v = 'test-value';
    const popover = wrapper.find(Popover);
    const list = popover.prop('content');
    list.props.setValue(v);
    expect(props.setValue.mock.calls).toEqual([[v]]);
  });

  it('hides the popover when the filter calls cancel', () => {
    wrapper.setState({ popoverVisible: true });
    const popover = wrapper.find(Popover);
    const list = popover.prop('content');
    list.props.cancel();
    expect(wrapper.state('popoverVisible')).toBe(false);
  });

  it('hides the popover when clicking outside of the open popover', () => {
    let mouseWithin = false;
    wrapper.setState({ popoverVisible: true });
    wrapper.instance().listRef = {
      current: {
        focusInput: () => {},
        isMouseWithin: () => mouseWithin,
      },
    };
    wrapper.instance().onBodyClicked();
    expect(wrapper.state('popoverVisible')).toBe(false);

    wrapper.setState({ popoverVisible: true });
    mouseWithin = true;
    wrapper.instance().onBodyClicked();
    expect(wrapper.state('popoverVisible')).toBe(true);

    wrapper.instance().listRef = {};
    wrapper.instance().onBodyClicked();
    expect(wrapper.state('popoverVisible')).toBe(true);
  });

  it('controls the visibility of the popover', () => {
    expect(wrapper.state('popoverVisible')).toBe(false);
    const popover = wrapper.find(Popover);
    popover.prop('onVisibleChange')(true);
    expect(wrapper.state('popoverVisible')).toBe(true);
  });

  it('attempts to focus the filter input when the component updates', () => {
    const fn = jest.fn();
    wrapper.instance().listRef = {
      current: {
        focusInput: fn,
      },
    };
    wrapper.setState({ popoverVisible: true });
    expect(fn.mock.calls.length).toBe(1);
  });

  describe('clear', () => {
    const clearValue = jest.fn();

    beforeEach(() => {
      clearValue.mockClear();
      wrapper.setProps({ clearValue, required: undefined, value: 'foo' });
    });

    it('renders with clear icon when not required and with a value', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('clicking clear icon clears value when not required without opening popover', () => {
      const stopPropagation = jest.fn();
      wrapper.find(Icon).simulate('click', { stopPropagation });

      expect(clearValue).toHaveBeenCalled();
      expect(wrapper.state('popoverVisible')).toBe(false);
      expect(stopPropagation).toHaveBeenCalled();
    });

    it('throws Error when attempting to clear when required', () => {
      expect(new NameSelector(props).clearValue).toThrowError('Cannot clear value of required NameSelector');
    });
  });
});
