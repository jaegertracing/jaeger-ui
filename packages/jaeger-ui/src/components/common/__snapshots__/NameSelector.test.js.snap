// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NameSelector> clear renders with clear icon when not required and with a value 1`] = `
<Popover
  content={
    <FilteredList
      cancel={[Function]}
      options={
        Array [
          "a",
          "b",
          "c",
        ]
      }
      setValue={[Function]}
      value="foo"
    />
  }
  mouseEnterDelay={0.1}
  mouseLeaveDelay={0.1}
  onVisibleChange={[Function]}
  overlayClassName="NameSelector--overlay u-rm-popover-content-padding"
  overlayStyle={Object {}}
  placement="bottomLeft"
  prefixCls="ant-popover"
  transitionName="zoom-big"
  trigger="click"
  visible={false}
>
  <h2
    className="NameSelector"
  >
    <span
      className="NameSelector--label"
    >
      a-label
      :
    </span>
    <BreakableText
      className="NameSelector--value"
      text="foo"
      wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
    />
    <IoChevronDown
      className="NameSelector--chevron"
    />
    <Icon
      className="NameSelector--clearIcon"
      onClick={[Function]}
      type="close"
    />
  </h2>
</Popover>
`;

exports[`<NameSelector> renders with is-invalid when required and without a value 1`] = `
<Popover
  content={
    <FilteredList
      cancel={[Function]}
      options={
        Array [
          "a",
          "b",
          "c",
        ]
      }
      setValue={[Function]}
      value={null}
    />
  }
  mouseEnterDelay={0.1}
  mouseLeaveDelay={0.1}
  onVisibleChange={[Function]}
  overlayClassName="NameSelector--overlay u-rm-popover-content-padding"
  overlayStyle={Object {}}
  placement="bottomLeft"
  prefixCls="ant-popover"
  transitionName="zoom-big"
  trigger="click"
  visible={false}
>
  <h2
    className="NameSelector is-invalid"
  >
    <BreakableText
      className="NameSelector--value"
      text="This is the placeholder"
      wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
    />
    <IoChevronDown
      className="NameSelector--chevron"
    />
  </h2>
</Popover>
`;

exports[`<NameSelector> renders without exploding 1`] = `
<Popover
  content={
    <FilteredList
      cancel={[Function]}
      options={
        Array [
          "a",
          "b",
          "c",
        ]
      }
      setValue={[Function]}
      value={null}
    />
  }
  mouseEnterDelay={0.1}
  mouseLeaveDelay={0.1}
  onVisibleChange={[Function]}
  overlayClassName="NameSelector--overlay u-rm-popover-content-padding"
  overlayStyle={Object {}}
  placement="bottomLeft"
  prefixCls="ant-popover"
  transitionName="zoom-big"
  trigger="click"
  visible={false}
>
  <h2
    className="NameSelector is-invalid"
  >
    <BreakableText
      className="NameSelector--value"
      text="This is the placeholder"
      wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
    />
    <IoChevronDown
      className="NameSelector--chevron"
    />
  </h2>
</Popover>
`;

exports[`<NameSelector> renders without is-invalid when not required and without a value 1`] = `
<Popover
  content={
    <FilteredList
      cancel={[Function]}
      options={
        Array [
          "a",
          "b",
          "c",
        ]
      }
      setValue={[Function]}
      value={null}
    />
  }
  mouseEnterDelay={0.1}
  mouseLeaveDelay={0.1}
  onVisibleChange={[Function]}
  overlayClassName="NameSelector--overlay u-rm-popover-content-padding"
  overlayStyle={Object {}}
  placement="bottomLeft"
  prefixCls="ant-popover"
  transitionName="zoom-big"
  trigger="click"
  visible={false}
>
  <h2
    className="NameSelector"
  >
    <BreakableText
      className="NameSelector--value"
      text="This is the placeholder"
      wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
    />
    <IoChevronDown
      className="NameSelector--chevron"
    />
  </h2>
</Popover>
`;